go-update/src/main/java/org/reactome/release/goupdate/GoTermInstanceModifier.java

```diff
@@ -0,0 +1,366 @@
+		{
+			referrers = (Collection<GKInstance>)instance.getReferers(ReactomeJavaConstants.activity);
+		}
+		
+		return referrers;
+	}
+	
+	/**
+	 * If a GO Term has certain referrers, it is not deletable. The rules (from Peter D.) are:<br/><br/><br/>
+	 * IF an GO biological process term has NOT been used as a goBiologicalProcess slot value for any event instance in gk_central, the obsolete GO term instance can be deleted from gk_central.<br/><br/>
+	 * IF a GO cellular component term has NOT been used as a compartment slot value for any physical entity or event instance in gk_central, the obsolete GO term instance can be deleted from gk_central.<br/><br/>
+	 * IF a GO molecular function term has NOT been used as the activity slot value for any catalystActivity instance in gk_central, the obsolete GO term instance can be deleted from gk_central.
+	 * @param instance - an instance to check.
+	 * @return true or false, if <code>instance</code> is deleteable, as per the above rules.
+	 * @throws Exception
+	 */
+	public static boolean isGoTermDeleteable(GKInstance instance) throws Exception
+	{
+		boolean isDeletable = false;
+		
+		Collection<GKInstance> referrers = GoTermInstanceModifier.getReferrersForGoTerm(instance);
+		
+		if (referrers != null && referrers.size() > 0)
+		{
+			isDeletable = false;
+		}
+		else
+		{
+			isDeletable = true;
+		}
+		
+		return isDeletable;
+	}
+	
+	
+	/**
+	 * Deletes a GO term from the database.
+	 * @param goTerms - The list of GO terms from the file. Needed to get the alternate GO IDs for things that refer to the thing that's about to be deleted.
+	 * @param allGoInstances - ALL GO instances from the database.
+	 */
+	public void deleteGoInstance(Map<String, Map<String,Object>> goTerms, Map<String, List<GKInstance>> allGoInstances, StringBuffer deletionStringBuilder)
+	{
+		try
+		{
+			String goId = (String) this.goInstance.getAttributeValue(ReactomeJavaConstants.accession);
+			
+			adaptor.deleteInstance(this.goInstance);
+			deletionStringBuilder.append("Deleting GO instance: \"").append(this.goInstance.toString()).append("\" (GO:").append(goId).append(")\n");
+		}
+		catch (Exception e)
+		{
+			logger.error("Error occurred while trying to delete instance: \""+this.goInstance.toString()+"\": "+e.getMessage());
+			e.printStackTrace();
+		}
+	}
+	
+	/**
+	 * Updates the relationships between GO terms in the database.
+	 * @param allGoInstances - Map of all GO instances in the database.
+	 * @param goProps - The properties to update with.
+	 * @param relationshipKey - The key to use to look up the values  in goProps.
+	 * @param reactomeRelationshipName - The name of the relationship, can be one of "is_a", "has_part", "part_of", "component_of", "regulates", "positively_regulates", "negatively_regulates".
+	 */
+	public void updateRelationship(Map<String, List<GKInstance>> allGoInstances, Map<String, Object> goProps, String relationshipKey, String reactomeRelationshipName)
+	{
```

Comment:
This function is needed because of the changed status of GO terms, or because of how things in the data model have changed (such as regulations), or both, or neither?

Reply:
Based on what I read in the Perl code, the relationship fields for a GO term may change, so the GO terms in the database need to be updated to keep in sync with the changes from GO.